index.js

```diff
@@ -85,109 +81,138 @@ PersonaClient.prototype.validateToken = function (token, scope, overridingScope,
-                    // put this key in redis with an expire
-                    _this.redisClient.multi().set("access_token:" + cacheKey, 'OK').expire("access_token:" + cacheKey, 60).exec(function (err, results) {
-                        _this.debug("cache: " + JSON.stringify(err) + JSON.stringify(results));
-                    });
-                    _this.debug("Verification passed for token " + cacheKey + ", cached for 60s");
-                    next(null, VALIDATED_TYPES.PERSONA);
+    var headScopeThenVerify = function headScope(scope, callback) {
+        var options = {
+            hostname: this.config.persona_host,
+            port: this.config.persona_port,
+            path: this.config.persona_oauth_route + token + "?scope=" + scope,
+            method: "HEAD"
+        };
+
+        this.http.request(options, function onSuccess(response) {
+            switch(response.statusCode) {
+            case 204:
+                return callback(null, "ok");
+            case 401:
+                return callback(ERROR_TYPES.VALIDATION_FAILURE, null);
+            case 403:
+                if(scope === "su") {
+                    // We tried using su and can go no further
+                    return callback(ERROR_TYPES.INSUFFICIENT_SCOPE, null);
                 } else {
-                    if (usingScope && usingScope !== "su") {
-                        // try su, they are allowed to perform any operation
-                        _this.validateToken (token, "su", overridingScope, next);
-                    } else {
-                        _this.debug("Verification failed for token " + cacheKey + " with status code " + oauthResp.statusCode);
-                        if (oauthResp.statusCode === 403) {
-                            next(ERROR_TYPES.INSUFFICIENT_SCOPE, null);
-                        } else {
-                            next(ERROR_TYPES.VALIDATION_FAILURE, null);
-                        }
-                    }
+                    // Try again with su
+                    return headScopeThenVerify("su", callback);
                 }
-            }).on ("error", function (e) {
-                _this.error("OAuth::validateToken problem: " + e.message);
-                next(e.message, null);
-            }).end();
-        }
-    });
+                break;
+            default:
+                return callback(ERROR_TYPES.COMMUNICATION_ISSUE, null);
+            }
+        }).on("error", function onError(error) {
+            return callback(error, null);
+        }).end();
+    };
+    headScopeThenVerify = headScopeThenVerify.bind(this);
+
+    var verifyToken = function verifyToken(publicKey) {
+        // Make sure we cache the key if this function was called as the result of a HTTP request to fetch it
+        this.publicKey = publicKey;
+        var jwtConfig = {
+            algorithms: ["RS256"]
+        };
+
+        jwt.verify(token, publicKey, jwtConfig, function onVerify(error, decodedToken) {
```

Comment:
Persona server should manage this at token generation time (see talis/platform#138).